In [3]:
import pandas as pd
import numpy as np
from scipy import stats
import statistics

In [4]:
# The ages of the original Brady children
bradys = np.array([14, 12, 11, 10, 8, 6, 8])

print(bradys)
#print(type(bradys))

# Exercise 1:
# Mean, median, mode--by hand
mean_brady_age = sum(bradys) / len(bradys)
print(mean_brady_age)

print(np.sort(bradys))
median_brady_age = 10 # I eyeballed it.

mode_brady_age = 8 # I eyeballed it.

# Variance is sum of (difference between each value and mean, squared), divided by sample size 
# NOTE: (I originally subtracted 1, not considering that this is the entire population of Bradys and not just a sample. :)

# Standard deviation is the sqrt of variance
variance_bradys = sum(map(lambda x: (x - mean_brady_age)**2, bradys)) / (len(bradys)) # - 1) <--THIS IS NOT A SAMPLE.
std_dev_bradys = np.sqrt(variance_bradys)
print("Variance: {}\nStandard deviation: {}".format(variance_bradys,std_dev_bradys))

# Standard error is standard deviation divided by sqrt of sample size
std_err_bradys = std_dev_bradys / np.sqrt(len(bradys) - 1)
print ("Standard error: {}".format(std_err_bradys))

# Now run the functions to calculate these values, and see how your work compares

df_bradys = pd.DataFrame()
df_bradys['ages'] = bradys
np.median(bradys)
print("Median: {}\nMode: {}".format(statistics.median(bradys),stats.mode(bradys)[0][0]))
df_bradys.describe()

[14 12 11 10  8  6  8]
9.85714285714
[ 6  8  8 10 11 12 14]
Variance: 6.408163265306122
Standard deviation: 2.531435020952764
Standard error: 1.0334540197243194
Median: 10
Mode: 8


ages
count   7.000000
mean    9.857143
std     2.734262
min     6.000000
25%     8.000000
50%    10.000000
75%    11.500000
max    14.000000

### Exercise 2:
One estimate of central tendency and one estimate of variance--
I'd choose the mean and the standard deviation. The low std dev here shows that there aren't extreme outliers in the series, so the mean should be fairly representative (not skewed too far toward either min or max values)

### Exercise 3: Cindy is now 7. Recalculate. What changed?
* I expect the mean will go up slightly, variance and standard deviation should go down.
* Median and mode remain the same.

In [11]:

bradys = np.array([14, 12, 11, 10, 8, 7, 8])
df_bradys['ages_3'] = bradys
np.median(bradys)
variance_bradys = sum(map(lambda x: (x - mean_brady_age)**2, bradys)) / (len(bradys)) #- 1)
std_dev_bradys = np.sqrt(variance_bradys)

print("Variance: {}\nStandard deviation: {}".format(variance_bradys,std_dev_bradys))

# Standard error is standard deviation divided by sqrt of sample size
std_err_bradys = std_dev_bradys / np.sqrt(len(bradys)) # - 1)
print("Standard error: {}".format(std_err_bradys))
print("Median: {}\nMode: {}".format(statistics.median(bradys),stats.mode(bradys)[0][0]))
# NOTE: The standard deviation and standard error differ slightly from the values in the solution...
# 2.3299 and .9511897 vs. what I calculated. Why? Is this difference significant?
# If I calculate using the same methods as the solution, I get the same values.
#variance_bradys = np.var(bradys)
#std_dev_bradys = np.std(bradys)
#*****

df_bradys.describe()

Variance: 5.448979591836734
Standard deviation: 2.334304948338313
Standard error: 0.8822843396415219
Median: 10
Mode: 8


ages     ages_3
count   7.000000   7.000000
mean    9.857143  10.000000
std     2.734262   2.516611
min     6.000000   7.000000
25%     8.000000   8.000000
50%    10.000000  10.000000
75%    11.500000  11.500000
max    14.000000  14.000000

### Exercise 4: Cousin Oliver has been replaced by Baby Jessica. Recalculate. What changed?
* I expect mean will drop; variance, standard deviation, standard error will increase.
* Median remains the same (replaced a value below the median with another value that is below the median).
* There is no longer a mode because Cousin Oliver and Bobby were both 8, and Oliver is gone.

In [12]:
bradys = np.array([14, 12, 11, 10, 8, 7, 1])
df_bradys['ages_4'] = bradys
np.median(bradys)
variance_bradys = sum(map(lambda x: (x - mean_brady_age)**2, bradys)) / (len(bradys)) # - 1)
std_dev_bradys = np.sqrt(variance_bradys)

variance_bradys = np.var(bradys)
std_dev_bradys = np.std(bradys)
print("Variance: {}\nStandard deviation: {}".format(variance_bradys,std_dev_bradys))

# Standard error is standard deviation divided by sqrt of sample size
std_err_bradys = std_dev_bradys / np.sqrt(len(bradys) - 1)
print("Standard error: {}".format(std_err_bradys))

print("Median: {}\nMode(?): {}".format(statistics.median(bradys),stats.mode(bradys)[0][0]))
# ^^^This calculates mode as "1", but in fact there is no mode
df_bradys.describe()


# Again, variance and std. dev are off slightly, but I get the same values as the solution if I use the same methods.


Variance: 15.428571428571429
Standard deviation: 3.927922024247863
Standard error: 1.6035674514745466
Median: 10
Mode(?): 1


ages     ages_3     ages_4
count   7.000000   7.000000   7.000000
mean    9.857143  10.000000   9.000000
std     2.734262   2.516611   4.242641
min     6.000000   7.000000   1.000000
25%     8.000000   8.000000   7.500000
50%    10.000000  10.000000  10.000000
75%    11.500000  11.500000  11.500000
max    14.000000  14.000000  14.000000

#### (still Exercise 4) Does this change my choice of central tendency or variance estimation methods?
* The median is now closer to the mean from previous exercises (median = 10, previous means were 9.857 and 10.0).
* I see that the standard deviation took a jump because Jessica's age is much more of an outlier than Oliver's age was.


### Exercise 5:
    TV Guide 20% fans 
    Entertainment Weekly 23% fans 
    Pop Culture Today 17% fans 
    SciPhi Phanatic 5% fans

Based on these numbers, what percentage of adult Americans would you estimate were Brady Bunch fans on the 50th anniversary of the show?


* I'm interested to know how representative of "adult Americans" we consider polls of readers of four media- or entertainment-oriented magazines. Is the selection of magazines/subscribers already introducing bias? 
* I mean, I can run the numbers, but without a more general, random sampling of adult Americans, it's possible (probable?) that the results aren't truly representative. For example, it's likely that Sports Illustrated subscribers believe an event like the World Series or the Super Bowl carries a lot more significance than subscribers to, say, SciPhi Phanatic. :)
* Speaking of SciPhi Phanatic, they are a clear outlier here--and maybe their lower result provides balance against the probable Brady-friendly bias of the other three sources? This poll result is going to bump up the variance and standard deviation though, making it harder to estimate the actual percentage of Brady Bunch fans among American adults.

In [18]:
brady_fans = np.array([20, 23, 17, 5])
df_brady_fans = pd.DataFrame()
df_brady_fans['polls'] = brady_fans

mean_brady_fans = sum(brady_fans) / len(brady_fans)
print("Median: {}".format(np.median(brady_fans)))

variance_brady_fans = sum(map(lambda x: (x - mean_brady_fans)**2, brady_fans)) / (len(brady_fans) - 1)
std_dev_brady_fans = np.sqrt(variance_brady_fans)
print("Variance: {}\nStandard deviation: {}".format(variance_brady_fans,std_dev_brady_fans))

# Standard error is standard deviation divided by sqrt of sample size
std_err_brady_fans = std_dev_brady_fans / np.sqrt(len(brady_fans) - 1)
print("Standard error: {}".format(std_err_brady_fans))

df_brady_fans.describe()


Median: 18.5
Variance: 62.25
Standard deviation: 7.88986691902975
Standard error: 4.55521678957215


polls
count   4.000000
mean   16.250000
std     7.889867
min     5.000000
25%    14.000000
50%    18.500000
75%    20.750000
max    23.000000

### (after looking at the solution)
Oh. Oh yeah. We can choose to just drop the outlier and calculate with what's left. :)

In [22]:
brady_fans = np.array([20, 23, 17])
df_brady_fans = pd.DataFrame()
df_brady_fans['polls'] = brady_fans

mean_brady_fans = sum(brady_fans) / len(brady_fans)
print("Median: {}".format(np.median(brady_fans)))

#variance_brady_fans = sum(map(lambda x: (x - mean_brady_fans)**2, brady_fans)) / (len(brady_fans) - 1)
#std_dev_brady_fans = np.sqrt(variance_brady_fans)
variance_brady_fans = np.var(brady_fans)
std_dev_brady_fans = np.std(brady_fans)
print("Variance: {}\nStandard deviation: {}".format(variance_brady_fans,std_dev_brady_fans))

# Standard error is standard deviation divided by sqrt of sample size
std_err_brady_fans = std_dev_brady_fans / np.sqrt(len(brady_fans) - 1)
print("Standard error: {}".format(std_err_brady_fans))

df_brady_fans.describe()

Median: 20.0
Variance: 6.0
Standard deviation: 2.449489742783178
Standard error: 1.732050807568877


polls
count    3.0
mean    20.0
std      3.0
min     17.0
25%     18.5
50%     20.0
75%     21.5
max     23.0